yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -225,12 +243,74 @@ private[yarn] class YarnAllocator(
       logInfo(s"Will request $missing executor containers, each with ${resource.getVirtualCores} " +
         s"cores and ${resource.getMemory} MB memory including $memoryOverhead MB overhead")
 
-      for (i <- 0 until missing) {
-        val request = createContainerRequest(resource)
-        amClient.addContainerRequest(request)
-        val nodes = request.getNodes
-        val hostStr = if (nodes == null || nodes.isEmpty) "Any" else nodes.last
-        logInfo(s"Container request (host: $hostStr, capability: $resource)")
+      // Calculated the number of executors we expected to satisfy all the preferred locality tasks
+      val localityAwareTaskCores = localityAwarePendingTaskNum * CPUS_PER_TASK
+      val expectedLocalityAwareContainerNum =
+        (localityAwareTaskCores + resource.getVirtualCores - 1) / resource.getVirtualCores
+
+      // Get the all the existed and locality matched containers
+      val existedMatchedContainers = allocatedHostToContainersMap.filter { case (host, _) =>
+        preferredLocalityToCounts.contains(host)
+      }
+      val existedMatchedContainerNum = existedMatchedContainers.values.map(_.size).sum
+
+      // The number of containers to allocate, divided into two groups, one with node locality,
+      // and the other without locality preference.
+      var requiredLocalityFreeContainerNum: Int = 0
+      var requiredLocalityAwareContainerNum: Int = 0
+
+      if (expectedLocalityAwareContainerNum <= existedMatchedContainerNum) {
+        // If the current allocated executor can satisfy all the locality preferred tasks,
```

Comment:
Oh, my bad, sorry for missing this part, I will change the strategy accordingly :).